File: Assets/Script/Example/Sprite.lua
-- [yue]: Script/Example/Sprite.yue
local Sprite = Dora.Sprite -- 1
local Node = Dora.Node -- 1
local threadLoop = Dora.threadLoop -- 1
local App = Dora.App -- 1
local ImGui = Dora.ImGui -- 1
local Vec2 = Dora.Vec2 -- 1
local Size = Dora.Size -- 1
local math = _G.math -- 1
local sprite -- 3
do -- 3
	local _with_0 = Sprite("Image/logo.png") -- 3
	_with_0.scaleX = 0.5 -- 4
	_with_0.scaleY = 0.5 -- 4
	_with_0.showDebug = true -- 5
	sprite = _with_0 -- 3
end -- 3
do -- 7
	local _with_0 = Node() -- 7
	_with_0:onTapMoved(function(touch) -- 8
		if not touch.first then -- 9
			return -- 9
		end -- 9
		sprite.position = sprite.position + touch.delta -- 10
	end) -- 8
	_with_0:addChild(sprite) -- 11
end -- 7
local windowFlags = { -- 16
	"NoResize", -- 16
	"NoSavedSettings" -- 16
} -- 16
return threadLoop(function() -- 17
	local width -- 18
	width = App.visualSize.width -- 18
	ImGui.SetNextWindowPos(Vec2(width - 10, 10), "FirstUseEver", Vec2(1, 0)) -- 19
	ImGui.SetNextWindowSize(Vec2(240, 520), "FirstUseEver") -- 20
	return ImGui.Begin("Sprite", windowFlags, function() -- 21
		ImGui.Text("Sprite (Yuescript)") -- 22
		ImGui.BeginChild("SpriteSetting", Vec2(-1, -40), function() -- 23
			local z = sprite.z -- 24
			do -- 25
				local changed -- 25
				changed, z = ImGui.DragFloat("Z", z, 1, -1000, 1000, "%.2f") -- 25
				if changed then -- 25
					sprite.z = z -- 26
				end -- 25
			end -- 25
			local x, y -- 27
			do -- 27
				local _obj_0 = sprite.anchor -- 27
				x, y = _obj_0.x, _obj_0.y -- 27
			end -- 27
			do -- 28
				local changed -- 28
				changed, x, y = ImGui.DragFloat2("Anchor", x, y, 0.01, 0, 1, "%.2f") -- 28
				if changed then -- 28
					sprite.anchor = Vec2(x, y) -- 29
				end -- 28
			end -- 28
			local spriteW, height -- 30
			do -- 30
				local _obj_0 = sprite.size -- 30
				spriteW, height = _obj_0.width, _obj_0.height -- 30
			end -- 30
			do -- 31
				local changed -- 31
				changed, spriteW, height = ImGui.DragFloat2("Size", spriteW, height, 1, 0, 1500, "%.f") -- 31
				if changed then -- 31
					sprite.size = Size(spriteW, height) -- 32
				end -- 31
			end -- 31
			local scaleX, scaleY = sprite.scaleX, sprite.scaleY -- 33
			do -- 34
				local changed -- 34
				changed, scaleX, scaleY = ImGui.DragFloat2("Scale", scaleX, scaleY, 0.01, -2, 2, "%.2f") -- 34
				if changed then -- 34
					sprite.scaleX, sprite.scaleY = scaleX, scaleY -- 35
				end -- 34
			end -- 34
			ImGui.PushItemWidth(-60, function() -- 36
				local angle = sprite.angle -- 37
				local changed -- 38
				changed, angle = ImGui.DragInt("Angle", math.floor(angle), 1, -360, 360) -- 38
				if changed then -- 38
					sprite.angle = angle -- 39
				end -- 38
			end) -- 36
			ImGui.PushItemWidth(-60, function() -- 40
				local angleX = sprite.angleX -- 41
				local changed -- 42
				changed, angleX = ImGui.DragInt("AngleX", math.floor(angleX), 1, -360, 360) -- 42
				if changed then -- 42
					sprite.angleX = angleX -- 43
				end -- 42
			end) -- 40
			ImGui.PushItemWidth(-60, function() -- 44
				local angleY = sprite.angleY -- 45
				local changed -- 46
				changed, angleY = ImGui.DragInt("AngleY", math.floor(angleY), 1, -360, 360) -- 46
				if changed then -- 46
					sprite.angleY = angleY -- 47
				end -- 46
			end) -- 44
			local skewX, skewY = sprite.skewX, sprite.skewY -- 48
			do -- 49
				local changed -- 49
				changed, skewX, skewY = ImGui.DragInt2("Skew", math.floor(skewX), math.floor(skewY), 1, -360, 360) -- 49
				if changed then -- 49
					sprite.skewX, sprite.skewY = skewX, skewY -- 50
				end -- 49
			end -- 49
			ImGui.PushItemWidth(-70, function() -- 51
				local opacity = sprite.opacity -- 52
				local changed -- 53
				changed, opacity = ImGui.DragFloat("Opacity", opacity, 0.01, 0, 1, "%.2f") -- 53
				if changed then -- 53
					sprite.opacity = opacity -- 54
				end -- 53
			end) -- 51
			return ImGui.PushItemWidth(-1, function() -- 55
				local color3 = sprite.color3 -- 56
				ImGui.SetColorEditOptions({ -- 57
					"DisplayRGB" -- 57
				}) -- 57
				if ImGui.ColorEdit3("", color3) then -- 58
					sprite.color3 = color3 -- 59
				end -- 58
			end) -- 59
		end) -- 23
		if ImGui.Button("Reset", Vec2(140, 30)) then -- 60
			local _with_0 = sprite.parent -- 61
			_with_0:removeChild(sprite) -- 62
			do -- 63
				local _with_1 = Sprite("Image/logo.png") -- 63
				_with_1.scaleX = 0.5 -- 64
				_with_1.scaleY = 0.5 -- 64
				_with_1.showDebug = true -- 65
				sprite = _with_1 -- 63
			end -- 63
			_with_0:addChild(sprite) -- 66
			return _with_0 -- 61
		end -- 60
	end) -- 66
end) -- 66

File: Tools/dora-rust/dora-test/src/tests/sprite.rs
/* Copyright (c) 2016-2026 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

use dora_ssr::*;

pub fn test() {
	let create_sprite = || -> Option<Sprite> {
		let mut sprite = match Sprite::with_file("Image/logo.png") {
			Some(sprite) => sprite,
			None => return None,
		};
		sprite.set_scale_x(0.5);
		sprite.set_scale_y(0.5);
		sprite.set_show_debug(true);
		sprite.set_touch_enabled(true);
		let mut sp = sprite.clone();
		sprite.slot(Slot::TAP_MOVED, Box::new(move |stack| {
			if let Some(touch) = stack.pop_cast::<Touch>() {
				if !touch.is_first() {
					return;
				}
				sp.set_position(&(sp.get_position() + touch.get_delta()));
			}
		}));
		Some(sprite)
	};
	let Some(mut sprite) = create_sprite() else {
		return;
	};

	let windows_flags =
		ImGuiWindowFlag::NoSavedSettings |
		ImGuiWindowFlag::NoResize;
	let mut imgui_node = Node::new();
	imgui_node.schedule(Box::new(move |_| {
		let width = App::get_visual_size().width;
		ImGui::set_next_window_bg_alpha(0.35);
		ImGui::set_next_window_pos_opts(&Vec2::new(width - 10.0, 10.0), ImGuiCond::FirstUseEver, &Vec2::new(1.0, 0.0));
		ImGui::set_next_window_size_opts(&Vec2::new(240.0, 520.0), ImGuiCond::Always);
		ImGui::begin_opts("Sprite", windows_flags, || {
			ImGui::text("Sprite (Rust)");
			ImGui::begin_child_opts("SpriteSetting", &Vec2::new(-1.0, -40.0), BitFlags::default(), BitFlags::default(), || {
				let (changed, z) = ImGui::drag_float_ret_opts("Z", sprite.get_z(), 1.0, -1000.0, 1000.0, "%.2f", ImGuiSliderFlag::ALWAYS_CLAMP.into());
				if changed {
					sprite.set_z(z);
				}
				let Vec2{x, y} = sprite.get_anchor();
				let (changed, x, y) = ImGui::drag_float2_ret_opts("Anchor", x, y, 0.01, 0.0, 1.0, "%.2f", BitFlags::default());
				if changed {
					sprite.set_anchor(&Vec2::new(x, y));
				}
				let Size{width, height} = sprite.get_size();
				let (changed, width, height) = ImGui::drag_float2_ret_opts("Size", width, height, 1.0, 0.0, 1500.0, "%.f", BitFlags::default());
				if changed {
					sprite.set_size(&Size::new(width, height));
				}
				let (changed, scale_x, scale_y) = ImGui::drag_float2_ret_opts("Scale", sprite.get_scale_x(), sprite.get_scale_y(), 0.01, -2.0, 2.0, "%.2f", BitFlags::default());
				if changed {
					sprite.set_scale_x(scale_x);
					sprite.set_scale_y(scale_y);
				}
				ImGui::push_item_width(-60.0, || {
					let (changed, angle) = ImGui::drag_int_ret_opts("Angle", sprite.get_angle() as i32, 1.0, -360, 360, "%d", BitFlags::default());
					if changed {
						sprite.set_angle(angle as f32);
					}
					let (changed, angle_x) = ImGui::drag_int_ret_opts("AngleX", sprite.get_angle_x() as i32, 1.0, -360, 360, "%d", BitFlags::default());
					if changed {
						sprite.set_angle_x(angle_x as f32);
					}
					let (changed, angle_y) = ImGui::drag_int_ret_opts("AngleY", sprite.get_angle_y() as i32, 1.0, -360, 360, "%d", BitFlags::default());
					if changed {
						sprite.set_angle_y(angle_y as f32);
					}
					let (changed, skew_x, skew_y) = ImGui::drag_int2_ret_opts("Skew", sprite.get_skew_x() as i32, sprite.get_skew_y() as i32, 1.0, -360, 360, "%d", BitFlags::default());
					if changed {
						sprite.set_skew_x(skew_x as f32);
						sprite.set_skew_y(skew_y as f32);
					}
				});
				ImGui::push_item_width(-70.0, || {
					let (changed, opacity) = ImGui::drag_float_ret_opts("Opacity", sprite.get_opacity(), 0.01, 0.0, 1.0, "%.2f", ImGuiSliderFlag::ALWAYS_CLAMP.into());
					if changed {
						sprite.set_opacity(opacity);
					}
				});
				ImGui::push_item_width(-1.0, || {
					let color3 = sprite.get_color3();
					let (changed, color3) = ImGui::color_edit3_ret_opts("Color", &color3, ImGuiColorEditFlag::DisplayRGB.into());
					if changed {
						sprite.set_color3(&color3);
					}
				});
			});
			if ImGui::button("Reset", &Vec2::new(140.0, 30.0)) {
				sprite.remove_from_parent(true);
				if let Some(new_sp) = create_sprite() {
					sprite = new_sp;
				}
			}
		});
		false
	}));
}